File: python/samples/templates_demo.py
import asyncio
import json
import random

from trinsic.proto.services.verifiablecredentials.templates.v1 import (
    TemplateField,
    FieldType,
    CreateCredentialTemplateRequest,
    GetCredentialTemplateRequest,
    DeleteCredentialTemplateRequest,
    SearchCredentialTemplatesRequest,
    FieldOrdering,
    AppleWalletOptions,
    UpdateCredentialTemplateRequest,
    TemplateFieldPatch,
)
from trinsic.proto.services.verifiablecredentials.v1 import (
    IssueFromTemplateRequest,
)
from trinsic.trinsic_service import TrinsicService
from trinsic.trinsic_util import trinsic_config, set_eventloop_policy


async def templates_demo():
    trinsic_service = TrinsicService(server_config=trinsic_config())
    ecosystem = await trinsic_service.provider.create_ecosystem()

    print("Ecosystem Token", trinsic_service.service_options.auth_token)

    # create example template
    # createTemplate() {
    create_request = CreateCredentialTemplateRequest(
        name=f"An Example Credential-{random.Random().randint(0,100)}",
        title="Example Credential",
        description="A credential for Trinsic's SDK samples",
        allow_additional_fields=False,
        fields={
            "firstName": TemplateField(
                title="First Name", description="Given name of holder"
            ),
            "lastName": TemplateField(
                title="Last Name", description="Surname of holder", optional=True
            ),
            "age": TemplateField(
                title="Age", description="Age in years of holder", type=FieldType.NUMBER
            ),
        },
        field_ordering={
            "firstName": FieldOrdering(order=0, section="Name"),
            "lastName": FieldOrdering(order=1, section="Name"),
            "age": FieldOrdering(order=2, section="Miscellaneous"),
        },
        apple_wallet_options=AppleWalletOptions(
            primary_field="firstName",
            secondary_fields=["lastName"],
            auxiliary_fields=["age"],
        ),
    )
    create_response = await trinsic_service.template.create(request=create_request)
    # }
    template = create_response
    assert template is not None
    assert template.data is not None
    assert template.data.id is not None
    assert template.data.schema_uri is not None

    template_id = template.data.id

    # update template
    # updateTemplate() {
    update_request = UpdateCredentialTemplateRequest(
        id=template_id,
        title="New Title",
        description="New Description",
        fields={
            "firstName": TemplateFieldPatch(title="New title for firstName"),
            "lastName": TemplateFieldPatch(description="New description for lastName"),
        },
        field_ordering={
            "age": FieldOrdering(order=0, section="Misc"),
            "firstName": FieldOrdering(order=1, section="Full Name"),
            "lastName": FieldOrdering(order=2, section="Full Name"),
        },
        apple_wallet_options=AppleWalletOptions(
            primary_field="age", secondary_fields=["firstName", "lastName"]
        ),
    )
    update_response = await trinsic_service.template.update(request=update_request)
    # }

    # issue credential from this template
    # issueFromTemplate() {
    values = json.dumps({"firstName": "Jane", "lastName": "Doe", "age": "42"})
    issue_response = await trinsic_service.credential.issue_from_template(
        request=IssueFromTemplateRequest(
            template_id=template.data.id, values_json=values
        )
    )
    # }
    json_document = json.loads(issue_response.document_json)
    assert json_document is not None

    json_keys = list(json_document.keys())
    assert "id" in json_keys
    assert "credentialSubject" in json_keys

    # TODO - Fix this, it's not part of the other tests
    # insert_response = await trinsic_service.wallet.insert_item(
    #     request=InsertItemRequest(item_json=json.dumps(json_document))
    # )
    # item_id = insert_response.item_id
    # frame = {
    #     "@context": "https://www.w3.org/2018/credentials/v1",
    #     "type": ["VerifiableCredential"],
    # }
    #
    # # create proof from input document
    # proof = await trinsic_service.credential.create_proof(
    #     request=CreateProofRequest(
    #         document_json=issue_response.document_json,
    #         reveal_document_json=json.dumps(frame),
    #     )
    # )
    # verify_result = await trinsic_service.credential.verify_proof(
    #     request=VerifyProofRequest(proof_document_json=proof.proof_document_json)
    # )
    # assert verify_result.is_valid
    #
    # # create proof from item id
    # proof = await trinsic_service.credential.create_proof(
    #     request=CreateProofRequest(
    #         item_id=item_id, reveal_document_json=json.dumps(frame)
    #     )
    # )
    # verify_result = await trinsic_service.credential.verify_proof(
    #     request=VerifyProofRequest(proof_document_json=proof.proof_document_json)
    # )
    # assert verify_result.is_valid
    #
    # try:
    #     # checkCredentialStatus() {
    #     check_response = await trinsic_service.credential.check_status(
    #         request=CheckStatusRequest(credential_status_id="")
    #     )
    #     # }
    # except:
    #     pass  # This is expected
    #
    # try:
    #     # updateCredentialStatus() {
    #     update_response = await trinsic_service.credential.update_status(
    #         request=UpdateStatusRequest(credential_status_id="", revoked=True)
    #     )
    #     # }
    # except:
    #     pass  # This is expected

    # getCredentialTemplate() {
    get_template_response = await trinsic_service.template.get(
        request=GetCredentialTemplateRequest(id=template.data.id)
    )
    # }
    # searchCredentialTemplate() {
    search_template_response = await trinsic_service.template.search(
        request=SearchCredentialTemplatesRequest(query="SELECT * FROM c")
    )
    # }
    # try:
    #     # deleteCredentialTemplate() {
    #     delete_template_response = await trinsic_service.template.delete(
    #         request=DeleteCredentialTemplateRequest(id=template.data.id)
    #     )
    #     # }
    # except:
    #     # permission denied on deletion - whatever
    #     pass


if __name__ == "__main__":
    set_eventloop_policy()
    asyncio.run(templates_demo())

File: java/src/test/java/trinsic/TemplatesDemo.java
package trinsic;

import com.google.gson.Gson;

import java.io.IOException;
import java.util.HashMap;
import java.util.UUID;
import java.util.concurrent.ExecutionException;

import trinsic.okapi.DidException;
import trinsic.services.TrinsicService;
import trinsic.services.verifiablecredentials.templates.v1.*;
import trinsic.services.verifiablecredentials.v1.CheckStatusRequest;
import trinsic.services.verifiablecredentials.v1.IssueFromTemplateRequest;
import trinsic.services.verifiablecredentials.v1.UpdateStatusRequest;

public class TemplatesDemo {
    private static String myEcosystemIdOrName = "default";

    public static void main(String[] args)
        throws IOException, DidException, ExecutionException, InterruptedException {
        run();
    }

    public static void run()
        throws IOException, DidException, ExecutionException, InterruptedException {
        var trinsic = new TrinsicService(TrinsicUtilities.getTrinsicServiceOptions());
        var account = trinsic.account().loginAnonymous(myEcosystemIdOrName).get();
        trinsic.setAuthToken(account);

        // create example template
        // createTemplate() {
        var templateRequest =
            CreateCredentialTemplateRequest.newBuilder()
                // Set metadata
                .setName("An Example Credential" + UUID.randomUUID())
                .setTitle("Example Credential")
                .setDescription("A credential for Trinsic's SDK samples")
                .setAllowAdditionalFields(false)
                // Construct fields
                .putFields("firstName",
                    TemplateField.newBuilder()
                        .setTitle("First Name")
                        .setDescription("Given name of holder")
                        .build()
                ).putFields("lastName",
                    TemplateField.newBuilder()
                        .setTitle("Last Name")
                        .setDescription("Surname of holder")
                        .setOptional(true)
                        .build()
                ).putFields("age",
                    TemplateField.newBuilder()
                        .setTitle("Age")
                        .setDescription("Age in years of holder")
                        .setType(FieldType.NUMBER)
                        .build()
                )
                // Specify ordering and categorization of fields for rendering of credential
                .putFieldOrdering("firstName",
                    FieldOrdering.newBuilder()
                        .setOrder(0)
                        .setSection("Name")
                        .build()
                ).putFieldOrdering("lastName",
                    FieldOrdering.newBuilder()
                        .setOrder(1)
                        .setSection("Name")
                        .build()
                ).putFieldOrdering("age",
                    FieldOrdering.newBuilder()
                        .setOrder(2)
                        .setSection("Miscellaneous")
                        .build()
                )
                // Specify Apple Wallet configuration
                .setAppleWalletOptions(
                    AppleWalletOptions.newBuilder()
                        .setPrimaryField("firstName")
                        .addSecondaryFields("lastName")
                        .addAuxiliaryFields("age")
                        .build()
                ).build();
        var template = trinsic.template().create(templateRequest).get();
        // }

        var templateId = template.getData().getId();

        assert template != null;
        assert template.hasData();
        assert !template.getData().getId().isEmpty();
        assert !template.getData().getSchemaUri().isEmpty();

        // Update template
        // updateTemplate() {
        var updateRequest =
            UpdateCredentialTemplateRequest.newBuilder()
                .setId(templateId)
                // Specify new title and description for the template
                .setTitle("New Title")
                .setDescription("New Description")
                // Update title/description for a few fields
                .putFields("firstName",
                    TemplateFieldPatch.newBuilder()
                        .setTitle("New title for firstName")
                        .build()
                )
                .putFields("lastName",
                    TemplateFieldPatch.newBuilder()
                        .setDescription("New description for lastName")
                        .build()
                )
                // Specify new ordering / categorization for template fields
                .putFieldOrdering("age",
                    FieldOrdering.newBuilder()
                        .setOrder(0)
                        .setSection("Misc")
                        .build()
                ).putFieldOrdering("firstName",
                    FieldOrdering.newBuilder()
                        .setOrder(1)
                        .setSection("Full Name")
                        .build()
                ).putFieldOrdering("lastName",
                    FieldOrdering.newBuilder()
                        .setOrder(2)
                        .setSection("Full Name")
                        .build()
                )
                // Specify new Apple Wallet configuration
                .setAppleWalletOptions(
                    AppleWalletOptions.newBuilder()
                        .setPrimaryField("age")
                        .addSecondaryFields("firstName")
                        .addSecondaryFields("lastName")
                        .build()
                ).build();

        var updatedTemplate = trinsic.template().update(updateRequest).get();
        // }

        // issue credential from this template
        // issueFromTemplate() {
        var valuesMap = new HashMap<String, Object>();
        valuesMap.put("firstName", "Jane");
        valuesMap.put("lastName", "Doe");
        valuesMap.put("age", 42);
        var valuesJson = new Gson().toJson(valuesMap);
        var issueResponse =
            trinsic
                .credential()
                .issueFromTemplate(
                    IssueFromTemplateRequest.newBuilder()
                        .setTemplateId(templateId)
                        .setValuesJson(valuesJson)
                        .build())
                .get();
        // }
        assert issueResponse != null;
        assert !issueResponse.getDocumentJson().isEmpty();

        var jsonDocument = new Gson().fromJson(issueResponse.getDocumentJson(), HashMap.class);
        assert jsonDocument.containsKey("id");
        assert jsonDocument.containsKey("credentialSubject");

        // getCredentialTemplate() {
        var getResponse =
            trinsic.template().get(GetCredentialTemplateRequest.newBuilder().setId(templateId).build()).get();
        // }

        // searchCredentialTemplate() {
        var searchResponse =
            trinsic
                .template()
                .search(
                    SearchCredentialTemplatesRequest.newBuilder()
                        .setQuery("SELECT * FROM c WHERE c.id = '" + templateId + "'")
                        .build())
                .get();
        // }

        try {
            // checkCredentialStatus() {
            var checkStatusResponse =
                trinsic.credential().checkStatus(CheckStatusRequest.newBuilder().build()).get();
            // }
        } catch (Exception e) { // This is okay as an example
        }

        try {
            // updateCredentialStatus() {
            trinsic.credential().updateStatus(UpdateStatusRequest.newBuilder().build());
            // }
        } catch (Exception e) { // This is okay as an example
        }

        try {
            // deleteCredentialTemplate() {
            var deleteResponse =
                trinsic
                    .template()
                    .delete(DeleteCredentialTemplateRequest.newBuilder().setId(templateId).build())
                    .get();
            // }
        } catch (Exception e) { // This is okay as an example
        }
        trinsic.shutdown();
    }
}
